src/lua/rust_interop.rs

```diff
@@ -56,6 +57,37 @@ pub fn register_libraries(lua: &rlua::Lua) -> LuaResult<()> {
     Ok(())
 }
 
+/// This function behaves just like Lua's built-in type() function, but also recognises classes and
+/// returns special names for them.
+#[allow(unused)]
+fn type_override(_lua: &rlua::Lua, arg: rlua::Value) -> Result<String, rlua::Error> {
+    use rlua::Value;
+
+    // Lua's type() returns the result of lua_typename(), but rlua does not make that available to
+    // us, so write our own.
+    match arg {
+        Value::Nil => Ok("nil".to_owned()),
+        Value::Boolean(_) => Ok("boolean".to_owned()),
+        Value::LightUserData(_) => Ok("userdata".to_owned()),
+        Value::UserData(_) => Ok("userdata".to_owned()),
+        Value::Integer(_) => Ok("number".to_owned()),
+        Value::Number(_) => Ok("number".to_owned()),
+        Value::String(_) => Ok("string".to_owned()),
+        Value::Function(_) => Ok("function".to_owned()),
+        Value::Thread(_) => Ok("thread".to_owned()),
+        Value::Error(e) => Err(e),
+        Value::Table(t) => {
+            // Handle our own objects specially: Return the metatable's .__class.name if it exists,
+            // else return "table".
+            let unused_error = Err(rlua::Error::CoroutineInactive);
```

Comment:
Using a `match` here is a much cleaner way to handle this case:

```rust
match t.get_metatable() {
    None => Ok("table".to_owned()),
    Some(t) => t.raw_get::<_, rlua::Table>("__class")?.raw_get("name")
}
```

Reply:
I don't like the `?` in there. When a table does not have `__class` member, then `type()` should return `"table"` instead of causing a Lua error. My "mess" was written with the goal to have `"table"` only once, but if you want, I can amend in the following change:
```diff
diff --git a/src/lua/rust_interop.rs b/src/lua/rust_interop.rs
index 7b3fb53..af0f1bb 100644
--- a/src/lua/rust_interop.rs
+++ b/src/lua/rust_interop.rs
@@ -79,11 +79,12 @@ fn type_override(_lua: &rlua::Lua, arg: rlua::Value) -> Result<String, rlua::Err
         Value::Table(t) => {
             // Handle our own objects specially: Return the metatable's .__class.name if it exists,
             // else return "table".
-            let unused_error = Err(rlua::Error::CoroutineInactive);
-            t.get_metatable().map_or(unused_error,
-                                     |t| t.raw_get("__class"))
-                .and_then(|t: rlua::Table| t.raw_get("name"))
-                .or(Ok("table".to_owned()))
+            match t.get_metatable() {
+                None => Ok("table".to_owned()),
+                Some(t) => t.raw_get("__class")
+                    .and_then(|t: rlua::Table| t.raw_get("name"))
+                    .or(Ok("table".to_owned()))
+            }
         }
     }
 }
```

Reply:
Ah right good catch. 

Yeah that unused error is something that stood out to me and is the thing I wanted to avoid. Your patch looks good.

Don't forget to remove the line above that suppresses `unused`

Reply:
Also for the future: for unused variables you can use `_variable_name` to make the warning be suppressed.

Reply:
Well, it is not unused in that sense. If there is no metatable, then a `Result` with that error is created via `map_or` (so it *is* used), but then the `.or()` later replaces that error with `Ok("table".to_owned())`. So in this sense the error is unused, but it is not unused in the "prefix with `_`"-sense. Basically, I wanted to write `"table"` only once, so turned "no metatable" into some kind of error.
I'll amend the commit with the patch above later.

Reply:
Got it. Since I missed the case that you don't want to return early if the meta table has the wrong thing, the original is fine too. It can be kinda awkward doing these `Option` and `Result` conversions, so not a big deal